### ISSUE TITLE
Handle partial object definitions

### ISSUE THREAD

**Issue by thebritican** [jahewson/elm-graphql#1]
Check out the discussion and [Fedor's solution](https://groups.google.com/d/msg/elm-dev/et7lnsDrPYI/bZIXx1EoDAAJ) to dynamic fields on GraphQLObjects. Like we talked about at the hackathon, it'll produce a lot of output, but maybe there can be a clever optimization through the compiler.

After this we can look at handling queries rather than just the schema!

**Comment by jahewson** [comment]
I took a look at this and I see a couple of problems.

**1.** It doesn't solve any problems. There's no benefit to writing:

```elm
type alias User =
  { id : Schema.User.ID
  , name : Schema.User.Name
  }
```

instead of simply:

```elm
type alias User =
  { id : String
  , name : String
  }
```

These two snippets are functionally identical, the extra indirection only causes confusion.

**2.** It doesn't scale, as it can't be nested. Let's give the user some friends:

```elm
type alias User =
  { id : Schema.User.ID
  , name : Schema.User.Name
  , friends : List Schema.User
  }
```

Now it's impossible for the following valid query to satisfy the type we just defined, because we don't want nested 2nd order "friends" to be retrieved (friends of friends).

```graphql
user {
    name
    friends {
        name
    }
}
```

So any nested objects need to be decomposed into records, leaving ony leaf types (scalars and enums) which can have type aliases. But we already have those built into Elm.

In conclusion, decomposing a query's type into one large Elm record is the only approach that can actually work, as GraphQL queries don't re-use nestable "objects" - each selection set is free to select whatever fields it wishes. The only exception is fragments, but I need to look into them more closely.

**Comment by jahewson** [comment]
The good news is that this is now implemented in the latest version - I'm generating one large record type for each query.
